winforms/src/toga_winforms/widgets/base.py

```diff
@@ -15,9 +18,39 @@
 
 class Scalable:
     SCALE_DEFAULT_ROUNDING = ROUND_HALF_EVEN
+    _dpi_scale = None
 
-    def init_scale(self, native):
-        self.dpi_scale = native.CreateGraphics().DpiX / 96
+    @property
+    def dpi_scale(self):
+        if Scalable._dpi_scale is None:
+            self.update_scale()
+        return Scalable._dpi_scale
+
+    @dpi_scale.setter
+    def dpi_scale(self, value):
+        Scalable._dpi_scale = value
```

Comment:
Storing the scale as a class attribute will make it impossible for different windows to have different scales. But that's the whole point of the "per monitor" scaling feature.

This also raises another question - if you have different scale factors set on different monitors, does a window update correctly when it's moved between them? I assume you have multiple monitors to test this, since you created #1930.

I suspect this might require implementing the DpiChanged event rather than DisplaySettingsChanged.

Reply:
Also, where is this setter called from?

Reply:
`DpiChanged` event does not fire consistently when DPI settings are changed. As an example, test the following test script:
```
import clr
from ctypes import wintypes, windll, c_bool, c_void_p

clr.AddReference("System.Windows.Forms")
import System.Windows.Forms as WinForms
from Microsoft.Win32 import SystemEvents


class SampleApp(WinForms.Form):
    def __init__(self):
        super().__init__()
        self.Text = "Sample WinForms App"

        self.label = WinForms.Label()
        self.label.AutoSize = True
        self.label.Text = "Hello, DPI Scaling!"
        self.label.Parent = self

        # Register the DpiChanged event handler
        self.DpiChanged += self.OnDpiChanged
        SystemEvents.DisplaySettingsChanged += self.OnDisplaySettingsChanged

    def OnDisplaySettingsChanged(self, sender, e):
        print("OnDisplaySettingsChanged triggered")

    def OnDpiChanged(self, sender, e):
        print("OnDpiChanged triggered")


if __name__ == "__main__":
    app = WinForms.Application()
    windll.user32.SetProcessDpiAwarenessContext.restype = wintypes.BOOL
    windll.user32.SetProcessDpiAwarenessContext.argtypes = [c_void_p]
    print(
        f"DPI Awareness Setting was successful: {c_bool(windll.user32.SetProcessDpiAwarenessContext(-4)).value}"
    )
    app.EnableVisualStyles()
    app.SetCompatibleTextRenderingDefault(False)
    form = SampleApp()
    app.Run(form)
```
When the DPI settings are changed while the app is running, you will find that only `OnDisplaySettingsChanged` is consistently fired. I had earlier found somewhere in Dotnet source code that `DpiChanged` event is conditionally turned off from firing when DPI settings are changed.

>Also, where is this setter called from?
The setter was used for testing but I have removed it.

Reply:
> I had earlier found somewhere in Dotnet source code that `DpiChanged` event is conditionally turned off from firing when DPI settings are changed.

I see some mention of that [here](https://stackoverflow.com/questions/68301676), and the relevant .NET code is indeed very difficult to understand. They mention `DpiChangedBeforeParent` and `DpiChangedAfterParent` as possibly being more reliable. 

But let's continue with your current approach for now.